net/netns/netns_darwin.go

```diff
@@ -33,60 +35,154 @@ var bindToInterfaceByRouteEnv = envknob.RegisterBool("TS_BIND_TO_INTERFACE_BY_RO
 
 var errInterfaceStateInvalid = errors.New("interface state invalid")
 
-// controlLogf marks c as necessary to dial in a separate network namespace.
-//
-// It's intentionally the same signature as net.Dialer.Control
-// and net.ListenConfig.Control.
+// routeCache caches the results of interfaceIndexFor calls to avoid
+// spamming the AF_ROUTE socket. This is used for soft
+// isolation mode where we do many route lookups.
+type routeCacheEntry struct {
+	ifIndex int
+	err     error
+}
+
+var (
+	routeCache     *lru.Cache[string, routeCacheEntry]
+	routeCacheOnce sync.Once
+)
+
+func getRouteCache() *lru.Cache[string, routeCacheEntry] {
+	routeCacheOnce.Do(func() {
+		routeCache, _ = lru.New[string, routeCacheEntry](256)
+	})
+	return routeCache
+}
+
+// ClearRouteCache clears the route cache. This should be called by the
+// network monitor when a link changes occur.
+func ClearRouteCache() {
+	getRouteCache().Purge()
+}
+
+// isInterfaceCoderInterface can be swapped out in tests.
+var isInterfaceCoderInterface func(int) bool = isInterfaceCoderInterfaceDefault
+
+func isInterfaceCoderInterfaceDefault(idx int) bool {
+	_, tsif, err := interfaces.Coder()
+	return err == nil && tsif != nil && tsif.Index == idx
+}
+
+// controlLogf binds c to the default interface if it would otherwise
+// be bound to the Coder interface.
 func controlLogf(logf logger.Logf, netMon *netmon.Monitor, network, address string, c syscall.RawConn) error {
-	if isLocalhost(address) {
-		// Don't bind to an interface for localhost connections.
+	if !shouldBindToDefaultInterface(logf, netMon, address) {
 		return nil
 	}
 
-	if disableBindConnToInterface.Load() {
-		logf("netns_darwin: binding connection to interfaces disabled")
+	idx, err := getDefaultInterfaceIndex(logf, netMon)
+	if err != nil {
 		return nil
 	}
 
-	idx, err := getInterfaceIndex(logf, netMon, address)
+	return bindConnToInterface(c, network, address, idx, logf)
+}
+
+// parseAddrForRouting returns the IP address for the given address, or an invalid
+// address if the address is not specified.
+func parseAddrForRouting(address string) (netip.Addr, error) {
+	host, _, err := net.SplitHostPort(address)
 	if err != nil {
-		// callee logged
-		return nil
+		return netip.Addr{}, fmt.Errorf("invalid address %q: %w", address, err)
+	}
+	if host == "" {
+		// netip.ParseAddr("") will fail
+		return netip.Addr{}, nil
 	}
 
-	return bindConnToInterface(c, network, address, idx, logf)
+	addr, err := netip.ParseAddr(host)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("invalid address %q: %w", address, err)
+	}
+	if addr.Zone() != "" {
+		// Addresses with zones *can* be represented as a route lookup with extra
+		// effort, but we don't use or support them currently.
```

Comment:
This comment was true on windows but idk if it is on mac, not really a big deal though to just leave it here though